src/Storage/Device/Local.php

```diff
@@ -119,26 +111,121 @@ public function upload(string $source, string $path, int $chunk = 1, int $chunks
+     * @throws \Exception
+     */
+    public function uploadData(string $data, string $path, string $contentType, int $chunk = 1, int $chunks = 1, array &$metadata = []): int
+    {
+        $this->createDirectory(\dirname($path));
+
+        if ($chunks === 1) {
+            if (! \file_put_contents($path, $data)) {
+                throw new Exception('Can\'t write file '.$path);
             }
-            \unlink($tmp);
+
+            return $chunks;
+        }
+        $tmp = \dirname($path).DIRECTORY_SEPARATOR.'tmp_'.\basename($path).DIRECTORY_SEPARATOR.\basename($path).'_chunks.log';
+
+        $this->createDirectory(\dirname($tmp));
+        if (! file_put_contents($tmp, "$chunk\n", FILE_APPEND)) {
+            throw new Exception('Can\'t write chunk log '.$tmp);
+        }
+
+        $chunkLogs = file($tmp);
+        if (! $chunkLogs) {
+            throw new Exception('Unable to read chunk log '.$tmp);
+        }
+
+        $chunksReceived = count(file($tmp));
+
+        if (! \file_put_contents(dirname($tmp).DIRECTORY_SEPARATOR.pathinfo($path, PATHINFO_FILENAME).'.part.'.$chunk, $data)) {
+            throw new Exception('Failed to write chunk '.$chunk);
+        }
+
+        if ($chunks === $chunksReceived) {
+            $this->joinChunks($path, $chunks);
 
             return $chunksReceived;
         }
 
         return $chunksReceived;
     }
 
+    private function joinChunks(string $path, int $chunks)
+    {
+        $tmp = \dirname($path).DIRECTORY_SEPARATOR.'tmp_'.\basename($path).DIRECTORY_SEPARATOR.\basename($path).'_chunks.log';
+        for ($i = 1; $i <= $chunks; $i++) {
+            $part = dirname($tmp).DIRECTORY_SEPARATOR.pathinfo($path, PATHINFO_FILENAME).'.part.'.$i;
+            $data = file_get_contents($part);
+            if (! $data) {
+                throw new Exception('Failed to read chunk '.$part);
+            }
+
+            if (! file_put_contents($path, $data, FILE_APPEND)) {
+                throw new Exception('Failed to append chunk '.$part);
+            }
+            \unlink($part);
+        }
+        \unlink($tmp);
+        \rmdir(dirname($tmp));
+    }
+
+    /**
+     * Transfer
+     *
+     * @param  string  $path
+     * @param  string  $destination
+     * @param  Device  $device
+     * @return string
+     */
+    public function transfer(string $path, string $destination, Device $device): bool
+    {
+        if (! $this->exists($path)) {
+            throw new Exception('File Not Found');
+        }
+        $size = $this->getFileSize($path);
+        $contentType = $this->getFileMimeType($path);
+
+        if ($size <= $this->transferChunkSize) {
+            $source = $this->read($path);
+
+            return $device->write($destination, $source, $contentType);
+        }
+
+        $totalChunks = \ceil($size / $this->transferChunkSize);
+        $counter = 0;
+        $metadata = ['content_type' => $contentType];
+        for ($counter; $counter < $totalChunks; $counter++) {
```

Comment:
We could move the `$counter` inline with the for loop here

Reply:
done.